tests/std/tests/VSO_0180469_ptr_cat/test.cpp

```diff
@@ -212,49 +218,90 @@ void ptr_cat_test_cases() {
-    // Adding volatile anywhere should explode
-    test_case_Equal_memcmp_is_safe_helper<false, volatile Elem1*, Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, Elem1*, volatile Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, volatile Elem1*, volatile Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, const volatile Elem1*, Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, const Elem1*, volatile Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, const volatile Elem1*, volatile Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, volatile Elem1*, const Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, Elem1*, const volatile Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, volatile Elem1*, const volatile Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, const volatile Elem1*, const Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, const Elem1*, const volatile Elem2*, equal_to<>>();
-    test_case_Equal_memcmp_is_safe_helper<false, const volatile Elem1*, const volatile Elem2*, equal_to<>>();
-
-    // Iterators that are not pointers should explode
-    test_case_Equal_memcmp_is_safe_helper<false, typename list<Elem1>::iterator, typename list<Elem2>::iterator,
-        equal_to<>>();
+    test_case_Equal_memcmp_is_safe_comparator<Expected, Elem1, Elem2, equal_to<>>();
+#ifdef __cpp_lib_concepts
+    test_case_Equal_memcmp_is_safe_comparator<Expected, Elem1, Elem2, ranges::equal_to>();
+#endif // __cpp_lib_concepts
 
     // equal_to< some other T > should explode
-    test_case_Equal_memcmp_is_safe_helper<false, Elem1*, Elem2*, equal_to<list<int>>>();
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Elem1*, Elem2*, equal_to<list<int>>> == false);
     // Non equal_to comparison functions should explode
     auto lambda = [](Elem1*, Elem2*) { return false; };
-    test_case_Equal_memcmp_is_safe_helper<false, Elem1*, Elem2*, decltype(lambda)>();
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Elem1*, Elem2*, decltype(lambda)> == false);
     // equal_to<T> should not explode
-    test_case_Equal_memcmp_is_safe_helper<Expected && is_same_v<Elem1, Elem2>, Elem1*, Elem2*, equal_to<Elem1>>();
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Elem1*, Elem2*, equal_to<Elem1>> == (Expected && is_same_v<Elem1, Elem2>) );
     // But again, not volatile
-    test_case_Equal_memcmp_is_safe_helper<false, Elem1*, Elem2*, equal_to<volatile Elem1>>();
-    test_case_Equal_memcmp_is_safe_helper<false, Elem1*, Elem2*, equal_to<const volatile Elem1>>();
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Elem1*, Elem2*, equal_to<volatile Elem1>> == false);
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Elem1*, Elem2*, equal_to<const volatile Elem1>> == false);
 }
 
+template <bool Expected, class Elem1, class Elem2>
 void equal_safe_test_cases() {

```

Comment:
Surprisingly, this templating is completely damaging the test. This is a compile-only test, and it's relying on non-template functions like `equal_safe_test_cases()` to instantiate the templates like `test_case_Equal_memcmp_is_safe<Expected, Elem1, Elem2>()`. This happens even though nothing attempts to call `equal_safe_test_cases()`. Adding this template prevents all of the instantiation, disabling all of the test coverage (!).

Reply:
So all the tests are green? Ship it!

Reply:
Enabling your enhanced test coverage revealed that I misunderstood how `contiguous_iterator` works with top-level const iterators (it doesn't). I fixed the product code accordingly.